### PR TITLE
rayjob controller: update EndTime to always be the time when the job deployment transitions to Complete status

### PR DESCRIPTION
**Pull request by andrewsykim**
## Why are these changes needed?

See #1849 

## Checks

- [X] I've made sure the tests are passing. 
- Testing Strategy
   - [X] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by andrewsykim** [comment]
cc @kevin85421

**Review by kevin85421 — COMMENTED** [review]
This PR differs somewhat from what I had in mind. Let me provide more details below. Feel free to let me know if it is still unclear.

The `EndTime` must be set when a RayJob transitions to `Complete`. If it is not set, [this line](https://github.com/ray-project/kuberay/blob/f2d94ffe213dd8f69481b09c474047cb899fa73b/ray-operator/controllers/ray/rayjob_controller.go#L263) will dereference a nil pointer. Currently, there are two possible cases that make the RayJob transitions to `Complete`:
* The JobStatus is terminal.
* The submitter K8s Job fails ([code](https://github.com/ray-project/kuberay/blob/f2d94ffe213dd8f69481b09c474047cb899fa73b/ray-operator/controllers/ray/rayjob_controller.go#L621)).

My thought is that (1) remove the assignment of `EndTime` in the above two cases (2) only set `EndTime` when the RayJob transitions to `Complete` ([code](https://github.com/ray-project/kuberay/blob/f2d94ffe213dd8f69481b09c474047cb899fa73b/ray-operator/controllers/ray/rayjob_controller.go#L519)).

**Comment by andrewsykim** [comment]
@kevin85421 makes sense, I updated the implementation. PTAL

File: ray-operator/apis/ray/v1/rayjob_types.go [hunk 1]
```diff
@@ -95,7 +95,8 @@ type RayJobStatus struct {
 	// It is not guaranteed to be set in happens-before order across separate operations.
 	// It is represented in RFC3339 form
 	StartTime *metav1.Time `json:"startTime,omitempty"`
-	// Represents time when the job was ended.
+	// EndTime is the time when status.jobDeploymentStatus transitioned to 'Complete' status.
```
Comment (kevin85421) [inline ray-operator/apis/ray/v1/rayjob_types.go:98]:
```suggestion
	// EndTime is the time when JobDeploymentStatus transitioned to 'Complete' status.
```

File: ray-operator/controllers/ray/rayjob_controller_test.go [hunk 2]
```diff
@@ -274,6 +274,21 @@ var _ = Context("Inside the default namespace", func() {
...
+			getJobInfo := func(context.Context, string) (*utils.RayJobInfo, error) {
+				return &utils.RayJobInfo{JobStatus: rayv1.JobStatusSucceeded}, nil
+			}
+			fakeRayDashboardClient.GetJobInfoMock.Store(&getJobInfo)
+			defer fakeRayDashboardClient.GetJobInfoMock.Store(nil)
+
+			Eventually(
+				getRayJobDeploymentStatus(ctx, myRayJob),
+				time.Second*15, time.Millisecond*500).Should(Equal(rayv1.JobDeploymentStatusComplete), "jobDeploymentStatus = %v", myRayJob.Status.JobDeploymentStatus)
+			Expect(myRayJob.Status.EndTime.Time).Should(BeTemporally(">", now), "EndTime = %v, Now = %v", myRayJob.Status.EndTime, now)
```
Comment (kevin85421) [inline ray-operator/controllers/ray/rayjob_controller_test.go:290]: Can we use something like `myRayJob.Status.EndTime.After(now)`?

File: ray-operator/apis/ray/v1/rayjob_types.go [hunk 3]
```diff
@@ -95,7 +95,8 @@ type RayJobStatus struct {
...
+	// This occurs when the Ray job reaches a terminal state (SUCCEEDED, FAILED, STOPPED)
```
Comment (kevin85421) [inline ray-operator/apis/ray/v1/rayjob_types.go:99]: There are two possibilities to transition to Complete at this moment. See https://github.com/ray-project/kuberay/pull/1872#pullrequestreview-1845027314 for more details.

Comment (andrewsykim) [reply ray-operator/apis/ray/v1/rayjob_types.go:99]: > There are two possibilities to transition to Complete at this moment

Right, but both possibilities still result in the Ray job reaching a terminal state right? Unless I'm missing something

File: ray-operator/controllers/ray/rayjob_controller_test.go (hunk 2 above)
Comment (andrewsykim) [reply ray-operator/controllers/ray/rayjob_controller_test.go:290]: sure, I agree this is cleaner than the "BeTemporally" checks

File: ray-operator/apis/ray/v1/rayjob_types.go (hunk 3 above)
Comment (andrewsykim) [reply ray-operator/apis/ray/v1/rayjob_types.go:99]: Ah, I see what you mean. The second condition when submitter job fails, is irrelevant to the Ray job status

Comment (andrewsykim) [reply ray-operator/apis/ray/v1/rayjob_types.go:99]: I updated the comments to also include when submitter job fails

File: ray-operator/apis/ray/v1/rayjob_types.go (hunk 1 above)
Comment (andrewsykim) [reply ray-operator/apis/ray/v1/rayjob_types.go]: done

**Review by kevin85421 — APPROVED** [review]

**Comment by andrewsykim** [comment]
Fixes https://github.com/ray-project/kuberay/issues/1849

**Pull request merged**